docs/source/tutorial/introduction.rst

```diff
@@ -0,0 +1,438 @@
+payloads to find those objects in the graph. For example, one approach you can
+take is to store the index as an attribute on the object you add to the graph:
+
+.. jupyter-execute::
+
+    class GraphNode:
+
+        def __init__(self, value):
+            self.value = value
+            self.index = None
+
+    graph = rx.PyGraph()
+    index = graph.add_node(GraphNode("A"))
+    graph[index].index = index
+
+Additionally, at any time you can find the index mapping to the data payload
+and build a mapping or update a reference to it. For example, building on the
+above example you can update the index references all at once after creation:
+
+.. jupyter-execute::
+
+    class GraphNode:
+        def __init__(self, value):
+            self.index = None
+            self.value = value
+
+        def __str__(self):
+            return f"GraphNode: {self.value} @ index: {self.index}"
+
+    class GraphEdge:
+        def __init__(self, value):
+            self.index = None
+            self.value = value
+
+        def __str__(self):
+            return f"EdgeNode: {self.value} @ index: {self.index}"
+
+    graph = rx.PyGraph()
+    graph.add_nodes_from([GraphNode(i) for i in range(5)])
+    graph.add_edges_from([(i, i + 1, GraphEdge(i)) for i in range(4)])
+    # Populate index attribute in GraphNode objects
+    for index in graph.node_indices():
+        graph[index].index = index
+    # Populate index attribute in GraphEdge objects
+    for index, data in graph.edge_index_map().items():
+        data[2].index = index
+    print("Nodes:")
+    for node in graph.nodes():
+        print(node)
+    print("Edges:")
+    for edge in graph.edges():
+        print(edge)
+
+Accessing edges and neighbors
+=============================
+
+You can access edges from a node using the :meth:`~rx.PyGraph.incident_edges`
+method:
+
+.. jupyter-execute::
+
+    print(G.incident_edges(2))
+
+which will return the edge indices of the edges incident a given node. You
+can also find the neighbor nodes using the :meth:`~rx.PyGraph.neighbors`
+method:
+
+.. jupyter-execute::
+
+    print(G.neighbors(2))
+
+which returns the node indices of any neighbors of a given node.
+
+Directed Graphs
+===============
+
+A directed graph is a graph that is made up of a set of nodes connected by
+directed edges (often called arcs). Edges have a directionality which is
+different from undirected graphs where edges have no notion of a direction to
+them. In retworkx the :class:`~retworkx.PyDiGraph` class is used to create
+directed graphs. For example:
+
+.. jupyter-execute::
+
+    import retworkx as rx
+    from retworkx.visualization import mpl_draw
+
+    path_graph = rx.generators.directed_path_graph(5)
+    mpl_draw(path_graph)
+
+In this example we created a 5 node directed path graph. This shows the
+directionality of the edges in the graph visualization with the arrow head
+pointing to the target node.
+
+Multigraphs
```

Comment:
There are some functions which do not work correctly on multigraphs, yet give no warning when a multigraph is passed to them. What is the long-term plan about this?

Will all functions eventually verify their input? Or will this be the responsibility of the user (for performance reasons)?  Either way, it would be useful to mention which choice you are going with.

Reply:
We should be documenting it in all the docstrings for those functions where it's not supported (for example https://qiskit.org/documentation/retworkx/stubs/retworkx.transitivity.html). If a note like that this is missing from a function that doesn't support a multigraphs that's a doc bug we should fix.

I think either approach is valid and we can handle this on a case by case basis as in some cases the performance overhead of validating an input graph has no parallel edges might be too high. But I agree we should be documenting it explicitly everywhere.

Reply:
In igraph, we are planning to start caching whether the graph has multiple edges, as well as a few other simple properties that are necessary for validating inputs. This is not implemented yet though, and will require some major changes.